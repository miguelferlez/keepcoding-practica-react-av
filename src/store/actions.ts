import { AxiosError } from "axios";
import type { AppThunk } from "./index";
import type { Advert } from "@/pages/adverts/types";
import type { Credentials } from "@/pages/auth/types";

//#region Auth

type AuthLoginFulfilled = {
  type: "auth/login/fulfilled";
};
type AuthLoginPending = {
  type: "auth/login/pending";
};
type AuthLoginRejected = {
  type: "auth/login/rejected";
  payload: AxiosError<{ message: string }>;
};
type AuthLogout = {
  type: "auth/logout";
};

const authLoginFulfilled = (): AuthLoginFulfilled => ({
  type: "auth/login/fulfilled",
});
const authLoginPending = (): AuthLoginPending => ({
  type: "auth/login/pending",
});
const authLoginRejected = (
  error: AxiosError<{ message: string }>,
): AuthLoginRejected => ({
  type: "auth/login/rejected",
  payload: error,
});

export function authLogin(
  credentials: Credentials,
  remember: boolean,
): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api, router }) {
    dispatch(authLoginPending());

    try {
      await api.auth.login(credentials, remember);
      dispatch(authLoginFulfilled());
      router.navigate(router.state.location.state?.from ?? "/", {
        replace: true,
      });
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(authLoginRejected(error));
      }
      throw error;
    }
  };
}

export function authLogout(): AuthLogout {
  return { type: "auth/logout" };
}

//#endregion

//#region Ui

type UiResetError = {
  type: "ui/reset-error";
};

export const uiResetError = (): UiResetError => {
  return { type: "ui/reset-error" };
};

//#endregion

//#region Adverts

type AdvertsLoadedFulfilled = {
  type: "adverts/loaded/fulfilled";
  payload: Advert[];
};
type AdvertsLoadedPending = {
  type: "adverts/loaded/pending";
};
type AdvertsLoadedRejected = {
  type: "adverts/loaded/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsLoadedFulfilled = (adverts: Advert[]): AdvertsLoadedFulfilled => ({
  type: "adverts/loaded/fulfilled",
  payload: adverts,
});
const advertsLoadedPending = (): AdvertsLoadedPending => ({
  type: "adverts/loaded/pending",
});
const advertsLoadedRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsLoadedRejected => ({
  type: "adverts/loaded/rejected",
  payload: error,
});

export function advertsLoaded(): AppThunk<Promise<void>> {
  return async function (dispatch, getState, { api }) {
    const state = getState();

    if (state.adverts.loaded) {
      return;
    }

    try {
      dispatch(advertsLoadedPending());
      const adverts = await api.adverts.getAdverts();
      setTimeout(() => {
        dispatch(advertsLoadedFulfilled(adverts));
      }, 1000);
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsLoadedRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

//#region Adverts tags

type AdvertsTagsFulfilled = {
  type: "adverts/tags/fulfilled";
  payload: string[];
};

type AdvertsTagsPending = {
  type: "adverts/tags/pending";
};

type AdvertsTagsRejected = {
  type: "adverts/tags/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsTagsFulfilled = (tags: string[]): AdvertsTagsFulfilled => ({
  type: "adverts/tags/fulfilled",
  payload: tags,
});

const advertsTagsPending = (): AdvertsTagsPending => ({
  type: "adverts/tags/pending",
});

const advertsTagsRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsTagsRejected => ({
  type: "adverts/tags/rejected",
  payload: error,
});

export function advertsTags(): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api }) {
    dispatch(advertsTagsPending());

    try {
      const tags = await api.adverts.getAdvertsTags();
      dispatch(advertsTagsFulfilled(tags));
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsTagsRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

//#region Adverts detail

type AdvertsDetailFulfilled = {
  type: "adverts/detail/fulfilled";
  payload: Advert;
};
type AdvertsDetailPending = {
  type: "adverts/detail/pending";
};
type AdvertsDetailRejected = {
  type: "adverts/detail/rejected";
  payload: AxiosError<{ message: string }>;
};

const advertsDetailFulfilled = (advert: Advert): AdvertsDetailFulfilled => ({
  type: "adverts/detail/fulfilled",
  payload: advert,
});
const advertsDetailPending = (): AdvertsDetailPending => ({
  type: "adverts/detail/pending",
});
const advertsDetailRejected = (
  error: AxiosError<{ message: string }>,
): AdvertsDetailRejected => ({
  type: "adverts/detail/rejected",
  payload: error,
});

export function advertsDetail(advertId: string): AppThunk<Promise<void>> {
  return async function (dispatch, _getState, { api }) {
    dispatch(advertsDetailPending());

    try {
      const advert = await api.adverts.getAdvertById(advertId);
      dispatch(advertsDetailFulfilled(advert));
    } catch (error) {
      if (error instanceof AxiosError) {
        dispatch(advertsDetailRejected(error));
      }
      throw error;
    }
  };
}

//#endregion

export type Actions =
  | AuthLoginFulfilled
  | AuthLoginPending
  | AuthLoginRejected
  | AuthLogout
  | UiResetError
  | AdvertsLoadedFulfilled
  | AdvertsLoadedPending
  | AdvertsLoadedRejected
  | AdvertsTagsFulfilled
  | AdvertsTagsPending
  | AdvertsTagsRejected
  | AdvertsDetailFulfilled
  | AdvertsDetailPending
  | AdvertsDetailRejected;
