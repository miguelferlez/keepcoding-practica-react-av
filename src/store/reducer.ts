import type { AxiosError } from "axios";
import type { Actions } from "./actions";
import type { Advert } from "../pages/adverts/types";

export type State = {
  auth: boolean;
  ui: {
    pending: boolean;
    error: AxiosError<{ message: string }> | null;
  };
  adverts: { loaded: boolean; data: Advert[] };
};

const defaultState: State = {
  auth: false,
  ui: {
    pending: false,
    error: null,
  },
  adverts: { loaded: false, data: [] },
};

export function auth(
  state = defaultState.auth,
  action: Actions,
): State["auth"] {
  switch (action.type) {
    case "auth/login/fulfilled":
      return true;
    case "auth/logout":
      return false;
    default:
      return state;
  }
}

export function ui(state = defaultState.ui, action: Actions): State["ui"] {
  switch (action.type) {
    case "auth/login/pending":
      return { pending: true, error: null };
    case "auth/login/fulfilled":
      return { pending: false, error: null };
    case "auth/login/rejected":
      return { pending: false, error: action.payload };
    case "adverts/loaded/pending":
      return { pending: true, error: null };
    case "adverts/loaded/fulfilled":
      return { pending: false, error: null };
    case "adverts/loaded/rejected":
      return { pending: false, error: action.payload };
    case "ui/reset-error":
      return { ...state, error: null };
    default:
      return state;
  }
}

export function adverts(
  state = defaultState.adverts,
  action: Actions,
): State["adverts"] {
  switch (action.type) {
    case "adverts/loaded/fulfilled":
      return { loaded: true, data: action.payload };
    default:
      return state;
  }
}
