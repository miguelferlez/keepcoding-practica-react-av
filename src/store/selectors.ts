import type { RootState } from ".";

export const getIsLogged = (state: RootState) => state.auth;

export const getUi = (state: RootState) => state.ui;
