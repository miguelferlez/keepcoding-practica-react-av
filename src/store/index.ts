import { combineReducers, createStore, applyMiddleware } from "redux";
import * as reducers from "./reducer";
import { useDispatch, useSelector } from "react-redux";
import { composeWithDevTools } from "@redux-devtools/extension/";
import * as thunk from "redux-thunk";
import type { Actions } from "./actions";
import type { createBrowserRouter } from "react-router";
import * as auth from "../services/auth";
import * as adverts from "../services/adverts";

const rootReducer = combineReducers(reducers);
type Router = ReturnType<typeof createBrowserRouter>;
type ExtraArgument = {
  api: { auth: typeof auth; adverts: typeof adverts };
  router: Router;
};

export default function configureStore(
  preloadedState: Partial<reducers.State>,
  router: Router,
) {
  const store = createStore(
    rootReducer,
    preloadedState as never,
    composeWithDevTools(
      applyMiddleware(
        thunk.withExtraArgument<reducers.State, Actions, ExtraArgument>({
          api: { auth, adverts },
          router,
        }),
      ),
    ),
  );

  return store;
}

export type AppStore = ReturnType<typeof configureStore>;
export type AppGetState = AppStore["getState"];
export type RootState = ReturnType<AppGetState>;
export type AppDispatch = AppStore["dispatch"];
export type AppThunk<ReturnType = void> = thunk.ThunkAction<
  ReturnType,
  RootState,
  ExtraArgument,
  Actions
>;

export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
