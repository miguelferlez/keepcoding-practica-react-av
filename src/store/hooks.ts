import { useAppDispatch, useAppSelector } from ".";
import type { Credentials } from "../pages/auth/types";
import { authLogin, authLogout, uiResetError } from "./actions";
import { getIsLogged } from "./selectors";

export function useAuth() {
  return useAppSelector(getIsLogged);
}

export function useLoginAction() {
  const dispatch = useAppDispatch();
  return function (credentials: Credentials, remember: boolean) {
    return dispatch(authLogin(credentials, remember));
  };
}
export function useLogoutAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(authLogout());
  };
}

export function useUiResetErrorAction() {
  const dispatch = useAppDispatch();
  return function () {
    return dispatch(uiResetError());
  };
}
