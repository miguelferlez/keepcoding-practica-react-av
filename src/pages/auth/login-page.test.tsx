import { AxiosError } from "axios";
import { Provider } from "react-redux";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import LoginPage from "./login-page";
import type { RootState } from "@/store";
import { authLogin, uiResetError } from "@/store/actions";

vi.mock("@/store/actions");

describe("LoginPage", () => {
  const state: RootState = {
    auth: false,
    ui: {
      pending: { auth: false, adverts: false, tags: false },
      error: null,
    },
    adverts: { loaded: false, data: [], tags: [] },
  };
  const store = {
    getState: () => state,
    subscribe: () => {},
    dispatch: () => {},
  };
  const renderComponent = (error?: AxiosError<{ message: string }>) => {
    if (error) {
      state.ui.error = error;
    }

    return render(
      //@ts-expect-error: store configuration not needed
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
  };

  it("should render", () => {
    const { container } = renderComponent();

    expect(container).toMatchSnapshot();
  });

  it("should dispatch login action", async () => {
    renderComponent();
    const usernameInput = screen.getByLabelText(/username/i);
    const passwordInput = screen.getByLabelText(/password/i);
    const rememberCheck = screen.getByRole("checkbox");
    const submitButton = screen.getByRole("button");

    expect(submitButton).toHaveTextContent("Submit");
    expect(submitButton).toBeDisabled();
    await userEvent.type(usernameInput, "test@test.com");
    await userEvent.type(passwordInput, "1234");
    expect(submitButton).toBeEnabled();
    await userEvent.click(submitButton);
    expect(authLogin).toBeCalledWith(
      { email: "test@test.com", password: "1234" },
      false,
    );
    await userEvent.click(rememberCheck);
    expect(rememberCheck).toBeChecked();
    await userEvent.click(submitButton);
    expect(authLogin).toBeCalledWith(
      { email: "test@test.com", password: "1234" },
      true,
    );
  });

  it("should dispatch alert component with error message", async () => {
    const error = new AxiosError<{ message: string }>("Unauthorized");
    const { container } = renderComponent(error);
    const alert = screen.getByRole("alert");

    expect(container).toMatchSnapshot();
    expect(alert).toHaveTextContent(error.message);
    await userEvent.click(alert);
    expect(uiResetError).toHaveBeenCalled();
  });
});
