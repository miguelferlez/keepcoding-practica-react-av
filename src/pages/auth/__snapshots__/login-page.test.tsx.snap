// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should dispatch alert component with error message 1`] = `
<div>
  <div
    class="bg-background flex min-h-screen flex-col items-center justify-center"
  >
    <div
      class="xs:border-primary/15 xs:rounded-lg xs:border xs:shadow-xl p-8 sm:max-w-md"
    >
      <div
        class="flex flex-col gap-4"
      >
        <h2
          class="text-center"
        >
          Log In
        </h2>
        <form
          class="flex flex-col gap-2"
        >
          <div
            class="mb-4 w-full max-w-sm min-w-[200px]"
          >
            <div
              class="relative"
            >
              <input
                autocomplete="off"
                class="peer ease text-primary focus:border-primary w-full rounded-md border border-neutral-500/15 bg-transparent px-4 py-3 transition placeholder:text-transparent focus:shadow-sm focus:outline-none"
                id="email"
                name="email"
                placeholder="Username"
                required=""
                type="email"
                value=""
              />
              <label
                class="peer-focus:text-primary bg-background pointer-events-none absolute top-3.5 left-2.5 origin-left transform cursor-text px-1 text-neutral-500/30 transition-all peer-not-placeholder-shown:-top-2.5 peer-not-placeholder-shown:left-2.5 peer-not-placeholder-shown:scale-90 peer-focus:-top-1.5 peer-focus:left-2.5 peer-focus:scale-90 peer-focus:text-xs"
                for="email"
              >
                Username
                 *
              </label>
            </div>
          </div>
          <div
            class="mb-4 w-full max-w-sm min-w-[200px]"
          >
            <div
              class="relative"
            >
              <input
                autocomplete="off"
                class="peer ease text-primary focus:border-primary w-full rounded-md border border-neutral-500/15 bg-transparent px-4 py-3 transition placeholder:text-transparent focus:shadow-sm focus:outline-none"
                id="password"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <label
                class="peer-focus:text-primary bg-background pointer-events-none absolute top-3.5 left-2.5 origin-left transform cursor-text px-1 text-neutral-500/30 transition-all peer-not-placeholder-shown:-top-2.5 peer-not-placeholder-shown:left-2.5 peer-not-placeholder-shown:scale-90 peer-focus:-top-1.5 peer-focus:left-2.5 peer-focus:scale-90 peer-focus:text-xs"
                for="password"
              >
                Password
                 *
              </label>
            </div>
          </div>
          <div
            class="mb-4 flex items-center"
          >
            <div
              class="peer relative flex cursor-pointer items-center"
            >
              <input
                class="peer checked:border-primary checked:bg-primary h-5 w-5 cursor-pointer appearance-none rounded border border-neutral-500/15 transition-all checked:shadow-sm"
                id="remember"
                type="checkbox"
              />
              <span
                class="pointer-events-none absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 transform text-white opacity-0 peer-checked:opacity-100"
              >
                <svg
                  class="h-3.5 w-3.5"
                  fill="currentColor"
                  stroke="currentColor"
                  stroke-width="1"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
            <label
              class="peer-has-checked:text-primary -mb-1 ml-2 cursor-pointer text-sm text-neutral-500/30 transition-colors select-none"
              for="remember"
            >
              Remember me
            </label>
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            <span
              class="-mb-0.75"
            >
              Submit
            </span>
          </button>
        </form>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    >
      <div
        class="alert alert-error"
        role="alert"
      >
        <div
          class="flex items-center justify-between gap-1"
        >
          <div
            class="flex items-center justify-center"
          />
          <span
            class="-mb-1"
          >
            Unauthorized
          </span>
        </div>
        <button
          aria-label="Close"
          class="flex cursor-pointer items-center justify-center rounded-lg"
          type="button"
        >
          <span
            class="sr-only"
          >
            Close
          </span>
          <span
            class="material-symbols-outlined !text-xl !font-medium"
          >
            close
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`LoginPage > should render 1`] = `
<div>
  <div
    class="bg-background flex min-h-screen flex-col items-center justify-center"
  >
    <div
      class="xs:border-primary/15 xs:rounded-lg xs:border xs:shadow-xl p-8 sm:max-w-md"
    >
      <div
        class="flex flex-col gap-4"
      >
        <h2
          class="text-center"
        >
          Log In
        </h2>
        <form
          class="flex flex-col gap-2"
        >
          <div
            class="mb-4 w-full max-w-sm min-w-[200px]"
          >
            <div
              class="relative"
            >
              <input
                autocomplete="off"
                class="peer ease text-primary focus:border-primary w-full rounded-md border border-neutral-500/15 bg-transparent px-4 py-3 transition placeholder:text-transparent focus:shadow-sm focus:outline-none"
                id="email"
                name="email"
                placeholder="Username"
                required=""
                type="email"
                value=""
              />
              <label
                class="peer-focus:text-primary bg-background pointer-events-none absolute top-3.5 left-2.5 origin-left transform cursor-text px-1 text-neutral-500/30 transition-all peer-not-placeholder-shown:-top-2.5 peer-not-placeholder-shown:left-2.5 peer-not-placeholder-shown:scale-90 peer-focus:-top-1.5 peer-focus:left-2.5 peer-focus:scale-90 peer-focus:text-xs"
                for="email"
              >
                Username
                 *
              </label>
            </div>
          </div>
          <div
            class="mb-4 w-full max-w-sm min-w-[200px]"
          >
            <div
              class="relative"
            >
              <input
                autocomplete="off"
                class="peer ease text-primary focus:border-primary w-full rounded-md border border-neutral-500/15 bg-transparent px-4 py-3 transition placeholder:text-transparent focus:shadow-sm focus:outline-none"
                id="password"
                name="password"
                placeholder="Password"
                required=""
                type="password"
                value=""
              />
              <label
                class="peer-focus:text-primary bg-background pointer-events-none absolute top-3.5 left-2.5 origin-left transform cursor-text px-1 text-neutral-500/30 transition-all peer-not-placeholder-shown:-top-2.5 peer-not-placeholder-shown:left-2.5 peer-not-placeholder-shown:scale-90 peer-focus:-top-1.5 peer-focus:left-2.5 peer-focus:scale-90 peer-focus:text-xs"
                for="password"
              >
                Password
                 *
              </label>
            </div>
          </div>
          <div
            class="mb-4 flex items-center"
          >
            <div
              class="peer relative flex cursor-pointer items-center"
            >
              <input
                class="peer checked:border-primary checked:bg-primary h-5 w-5 cursor-pointer appearance-none rounded border border-neutral-500/15 transition-all checked:shadow-sm"
                id="remember"
                type="checkbox"
              />
              <span
                class="pointer-events-none absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 transform text-white opacity-0 peer-checked:opacity-100"
              >
                <svg
                  class="h-3.5 w-3.5"
                  fill="currentColor"
                  stroke="currentColor"
                  stroke-width="1"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M16.707 5.293a1 1 0 010 1.414l-8 8a1 1 0 01-1.414 0l-4-4a1 1 0 011.414-1.414L8 12.586l7.293-7.293a1 1 0 011.414 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
            <label
              class="peer-has-checked:text-primary -mb-1 ml-2 cursor-pointer text-sm text-neutral-500/30 transition-colors select-none"
              for="remember"
            >
              Remember me
            </label>
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            <span
              class="-mb-0.75"
            >
              Submit
            </span>
          </button>
        </form>
      </div>
    </div>
    <div
      class="fixed top-11.5 left-1/2 z-50 -translate-x-1/2"
    />
  </div>
</div>
`;
